### PR TITLE
Add neon optimization

### PR DESCRIPTION
**Pull request by aalmaksour**

### CODE REVIEW

File: src/arch/dotproductneon.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+namespace tesseract {
+    int32_t IntDotProductNEON(const int8_t* u, const int8_t* v, int n) {
+        fprintf(stderr, "IntDotProductNEON can't be used\n");
+        abort();
+    }
+}  // namespace tesseract
+
+#else  // !(defined(__arm__) || defined(__aarch64__))
+
+#include "dotproductneon.h"
```
Comment (ab-dragon) [inline src/arch/dotproductneon.cpp]: given that you are including this header in both the cases, you might want to move this include outside of #if-#else-#endif

File: src/arch/dotproductneon.cpp [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            int8x16_t packed1 = vld1q_s8(u);
+            int8x16_t packed2 = vld1q_s8(v);
+            sum_4_32 = madd_16_of_8b_to_4_32b(packed1, packed2);
+            while(offset <= max_offset) {
+                    packed1 = vld1q_s8(u + offset);
+                    packed2 = vld1q_s8(v + offset);
+                    offset += 16;
+                    int32x4_t sum_4_32_part = madd_16_of_8b_to_4_32b(packed1, packed2);
+                    sum_4_32 = vaddq_s32(sum_4_32, sum_4_32_part);
+            }
```
Comment (ab-dragon) [inline src/arch/dotproductneon.cpp:43]: you mentioned elsewhere that this function could be called tens of thousands times...I am wondering if loop unrolling will give you some extra performance boost. Here's one article that explains the concept...it is very simple really...reduce the number of time you do the comparison against the limit.
http://fastcpp.blogspot.com/2011/04/how-to-unroll-loop-in-c.html

Just a suggestion really if you want to give it a go.

File: src/arch/intsimdmatrixsse.cpp [hunk 3]
```diff
@@ -27,7 +27,7 @@ namespace tesseract {
 #ifdef __SSE4_1__
 // Computes part of matrix.vector v = Wu. Computes 1 result.
 static void PartialMatrixDotVector1(const int8_t* wi, const double* scales,
-                                    const int8_t* u, int num_in, int num_out,
+                                    const int8_t* u, int num_in, int /*num_out*/,
```
Comment (ab-dragon) [inline src/arch/intsimdmatrixsse.cpp:30]: is this change needed?

File: src/arch/dotproductneon.cpp (hunk 1 above)
Comment (aalmaksour) [reply src/arch/dotproductneon.cpp]: Done

File: src/arch/intsimdmatrixsse.cpp (hunk 3 above)
Comment (aalmaksour) [reply src/arch/intsimdmatrixsse.cpp:30]: The parameter is not used so I commented it out for more clarity.

File: src/arch/dotproductneon.cpp (hunk 2 above)
Comment (aalmaksour) [reply src/arch/dotproductneon.cpp:43]: We tried to unroll the loop to reduce comparison time and allow Neon to use more registers simultaneously. We did not get a significant time improvement.

**Review by ab-dragon — APPROVED** [review]
